File: src/backend/Intelibill.Domain/Entities/PurchaseOrder.cs
using Intelibill.Domain.Common;
using Intelibill.Domain.Enums;

namespace Intelibill.Domain.Entities;

public sealed class PurchaseOrder : BaseEntity
{
    private readonly List<PurchaseOrderLine> _lines = [];

    public Guid ShopId { get; private set; }
    public Guid? SupplierId { get; private set; }
    public string PurchaseOrderNumber { get; private set; } = string.Empty;
    public DateOnly? OrderDate { get; private set; }
    public DateOnly? ExpectedDeliveryDate { get; private set; }
    public string? SupplierReferenceNumber { get; private set; }
    public string? Notes { get; private set; }
    public PurchaseOrderStatus Status { get; private set; }
    public IReadOnlyList<PurchaseOrderLine> Lines => _lines.AsReadOnly();

    public decimal ExpectedTotal => _lines.Sum(l => l.LineTotal);

    private PurchaseOrder() { }

    public static PurchaseOrder CreateDraft(
        Guid shopId,
        string purchaseOrderNumber,
        Guid? supplierId,
        DateOnly? orderDate,
        DateOnly? expectedDeliveryDate,
        string? supplierReferenceNumber,
        string? notes)
    {
        return new PurchaseOrder
        {
            ShopId = shopId,
            SupplierId = supplierId,
            PurchaseOrderNumber = purchaseOrderNumber,
            OrderDate = orderDate,
            ExpectedDeliveryDate = expectedDeliveryDate,
            SupplierReferenceNumber = NormalizeOptional(supplierReferenceNumber),
            Notes = NormalizeOptional(notes),
            Status = PurchaseOrderStatus.Draft,
        };
    }

    public PurchaseOrderLine AddLine(Guid itemId, string description, int expectedQuantity, decimal unitCost)
    {
        var line = PurchaseOrderLine.Create(Id, itemId, description, expectedQuantity, unitCost);
        _lines.Add(line);
        return line;
    }

    public void UpdateDraft(
        Guid? supplierId,
        DateOnly? orderDate,
        DateOnly? expectedDeliveryDate,
        string? supplierReferenceNumber,
        string? notes,
        IReadOnlyList<(Guid ItemId, string Description, int ExpectedQuantity, decimal UnitCost)> lines)
    {
        if (Status != PurchaseOrderStatus.Draft)
            throw new InvalidOperationException("Only draft purchase orders can be updated.");

        SupplierId = supplierId;
        OrderDate = orderDate;
        ExpectedDeliveryDate = expectedDeliveryDate;
        SupplierReferenceNumber = NormalizeOptional(supplierReferenceNumber);
        Notes = NormalizeOptional(notes);
        _lines.Clear();
        foreach (var line in lines)
        {
            AddLine(line.ItemId, line.Description, line.ExpectedQuantity, line.UnitCost);
        }
    }

    private static string? NormalizeOptional(string? value)
    {
        if (string.IsNullOrWhiteSpace(value))
            return null;
        return value.Trim();
    }
}

File: src/mobile/android/intelibill_mobile/lib/src/features/purchase_orders/domain/entities/purchase_order.dart
import 'package:equatable/equatable.dart';
import 'package:intelibill_mobile/src/features/purchase_orders/domain/entities/purchase_order_line.dart';
import 'package:intelibill_mobile/src/features/purchase_orders/domain/entities/purchase_order_receipt.dart';
import 'package:intelibill_mobile/src/features/purchase_orders/domain/entities/purchase_order_status.dart';

class PurchaseOrder extends Equatable {
  const PurchaseOrder({
    required this.purchaseOrderId,
    required this.purchaseOrderNumber,
    required this.status,
    this.supplierId,
    this.orderDate,
    this.expectedDeliveryDate,
    this.supplierReferenceNumber,
    this.notes,
    required this.lines,
    required this.expectedTotal,
    required this.createdAt,
    this.supplierName,
    this.supplierReference,
    this.receivedQuantity = 0,
    this.cancellationReason,
    this.closedAt,
    this.closedBy,
    this.closeReason,
    this.receipts = const [],
  });

  final String purchaseOrderId;
  final String purchaseOrderNumber;
  final PurchaseOrderStatus status;
  final String? supplierId;
  final DateTime? orderDate;
  final DateTime? expectedDeliveryDate;
  final String? supplierReferenceNumber;
  final String? notes;
  final List<PurchaseOrderLine> lines;
  final double expectedTotal;
  final DateTime createdAt;
  final String? supplierName;
  final String? supplierReference;
  final int receivedQuantity;
  final String? cancellationReason;
  final DateTime? closedAt;
  final String? closedBy;
  final String? closeReason;
  final List<PurchaseOrderReceipt> receipts;

  int get expectedQuantity =>
      lines.fold(0, (sum, line) => sum + line.expectedQuantity);

  int get remainingQuantity =>
      lines.fold(0, (sum, line) => sum + line.remainingQuantity);

  @override
  List<Object?> get props => [
    purchaseOrderId,
    purchaseOrderNumber,
    status,
    supplierId,
    orderDate,
    expectedDeliveryDate,
    supplierReferenceNumber,
    notes,
    lines,
    expectedTotal,
    createdAt,
    supplierName,
    supplierReference,
    receivedQuantity,
    cancellationReason,
    closedAt,
    closedBy,
    closeReason,
    receipts,
  ];
}
